rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+# Reducing Primitive Creation Overhead for Dynamic Shapes
+
+## Motivation
+
+oneDNN API requires users to specify tensors dimensions to create a primitive,
+which makes the primitive dependent on the tensors's shapes. Primitive creation is
+an expensive operation, this is why oneDNN has a primitive cache to reduce the
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
```

Comment:
I think the points about RUNTIME_DIMENSION here are orthogonal to kernel cache.

RUNTIME_DIMENSION could actually be useful to increase kernel cache hit rate. When RUNTIME_DIMENSION is used on API side, we have extra information during dispatching that a dimension will change, This allows the dispatched implementations to make a decision:
- generate a kernel that is loosely dependent on the dimension that will vary in order to reduce overall kernel generation time (even if it loses some performance)
- generate a kernel that is tightly dependent on the dimension to improve performance of the primitive

Without that information the dispatched impl does not know if it can make the kernel tightly dependent on the varying dimension or not, and will have to statically chose between losing performance to potentially reduce overall kernel creation time, or get best performance.

Reply:
> generate a kernel that is loosely dependent on the dimension that will vary in order to reduce overall kernel generation time (even if it loses some performance)

@mgouicem, to some extent, I think this is a mistake as it introduces a lot more complexity. First, it creates another case where implementation `A` is worse than implementation `B`, but our heuristics call implementation `A`. Second, we now need to validate every input under multiple "views". Finally, there is a tradeoff between how many resources we should spend generating the important specializations in the RUNTIME_DIMENSION kernel and performance. This tradeoff is use case specific, creating a whole space of "optimal" solutions, from which we are only targeting one point.

To get around this, I personally think that a "RUNTIME DIMENSION" primitive should just map to a set of kernel specializations. These specializations should be exactly the set of possible kernels generated by standard oneDNN use, and creation should be performed lazily so users only pay for what is required. Under this structure, I don't think these are quite orthogonal (although they can be implemented in parallel). At its core, there are a two key operations: 
A dispatching heuristic mapping problem to specialized kernel (Mostly equivalent to pd creation)
Generating the specialized kernel (Mostly equivalent to primitive init)

Under this guise, the primitive cache is a tool to skip the dispatching heuristic and generating the kernel, the kernel cache is a tool to skip "Generating the specialized kernel". RUNTIME_DIMENSIONS is a tool to both "use an optimized dispatching heuristic" and skip "Generating the specialized kernel". Finally, there is another possible tool to just "use an optimized dispatching heuristic" that could be composed with the kernel cache as well.

Reply:
That comment about runtime_dimension support was independant of the presence of a kernel cache or not. It was mostly in the eventuality that we don't want / can't have a small subset of fixed kernels to cover all shapes with good performance.
Relying on runtime_dimension API would allow to have it both ways:
- rely on a large set of shape-specific kernels for best performance for static shapes
- rely on a small set of "shape-agnostic" kernels for best performance for dynamic shapes.

> To get around this, I personally think that a "RUNTIME DIMENSION" primitive should just map to a set of kernel specializations. These specializations should be exactly the set of possible kernels generated by standard oneDNN use, 

This is the big question mark to me. Does that set exist for every primitive? how big is it? And how much performance do we lose by making it smaller when too big?
Do we have clear indication that we can have such a small set without losing too much performance for the static shapes case?

Reply:
> RUNTIME_DIMENSION could actually be useful to increase kernel cache hit rate. 

In theory yes. In practice, I'm not sure that frameworks have enough information on their level to decide on whether they should create a primitive with runtime dimensions or not. In general, I find the feature to be not scalable and based on a workaround-like design. It was designed specifically for CPU matmul (even though it's not tied to it) assuming that we have a few GEMM kernels that can cover all the cases therefore the users can always create a matmul primitive with all dimensions to be runtime without any performance penalty. And this works fine for CPU matmul, but I have a doubt that it can be scaled up to cover other cases.

Reply:
As I know (please correct me if I am wrong) with profiling run mechanism in graph mode of PyTorch/IPEX, the dynamic dimensions can be detected. Actually oneDNN Graph has already received the request of support compilation with dynamic shape inputs. From the perspective of Graph API, frameworks tell which dimension will be dynamic and oneDNN Graph backend take those dynamic dimensions to generate kernels. Specifically for DNNL backend, it will create primitive with dynamic/runtime dimensions.